File: code/backend/engine/boardBinary.go
package engine

import (
	"digital-innovation/stratego/models"
	"encoding/base64"
	"fmt"
)

// Binary format: 8 bits per square
// Bit 7: Occupied | Bits 6-2: Piece type | Bit 1: Color | Bit 0: Moved
const (
	BitOccupied    = 1 << 7
	BitColor       = 1 << 1
	BitMoved       = 1 << 0
	MaskPieceType  = 0x7C
	ShiftPieceType = 2
)

const (
	PieceIDFlag       byte = 1
	PieceIDBomb       byte = 2
	PieceIDSpy        byte = 3
	PieceIDScout      byte = 4
	PieceIDMiner      byte = 5
	PieceIDSergeant   byte = 6
	PieceIDLieutenant byte = 7
	PieceIDCaptain    byte = 8
	PieceIDMajor      byte = 9
	PieceIDColonel    byte = 10
	PieceIDGeneral    byte = 11
	PieceIDMarshal    byte = 12
)

var rankToPieceID = map[byte]byte{
	'0': PieceIDFlag, 'B': PieceIDBomb, '1': PieceIDSpy, '2': PieceIDScout,
	'3': PieceIDMiner, '4': PieceIDSergeant, '5': PieceIDLieutenant, '6': PieceIDCaptain,
	'7': PieceIDMajor, '8': PieceIDColonel, '9': PieceIDGeneral, 'M': PieceIDMarshal,
}

var pieceIDToRank = func() [32]byte {
	var a [32]byte
	for k, v := range rankToPieceID {
		a[v] = k
	}
	return a
}()

var pieceTypeToID = map[*models.PieceType]byte{
	&models.Flag: PieceIDFlag, &models.Bomb: PieceIDBomb, &models.Spy: PieceIDSpy,
	&models.Scout: PieceIDScout, &models.Miner: PieceIDMiner, &models.Sergeant: PieceIDSergeant,
	&models.Lieutenant: PieceIDLieutenant, &models.Captain: PieceIDCaptain, &models.Major: PieceIDMajor,
	&models.Colonel: PieceIDColonel, &models.General: PieceIDGeneral, &models.Marshal: PieceIDMarshal,
}

var idToPieceType = map[byte]*models.PieceType{
	PieceIDFlag: &models.Flag, PieceIDBomb: &models.Bomb, PieceIDSpy: &models.Spy,
	PieceIDScout: &models.Scout, PieceIDMiner: &models.Miner, PieceIDSergeant: &models.Sergeant,
	PieceIDLieutenant: &models.Lieutenant, PieceIDCaptain: &models.Captain, PieceIDMajor: &models.Major,
	PieceIDColonel: &models.Colonel, PieceIDGeneral: &models.General, PieceIDMarshal: &models.Marshal,
}

// Encode full 10x10 board to 100 bytes
func EncodeBoard(board *Board, movedPositions map[Position]bool) []byte {
	data := make([]byte, 100)
	field := board.GetField()

	for y := range 10 {
		for x := range 10 {
			piece := field[y][x]
			if piece == nil {
				continue
			}

			cell := byte(BitOccupied)

			// Get piece ID from rank character
			rank := piece.GetType().GetRank()
			pieceID := rankToPieceID[rank]
			cell |= (pieceID << ShiftPieceType) & MaskPieceType

			if piece.GetOwner().GetID() == 2 {
				cell |= BitColor
			}

			if movedPositions != nil && movedPositions[NewPosition(x, y)] {
				cell |= BitMoved
			}

			data[y*10+x] = cell
		}
	}

	return data
}

// Decode 100 bytes to full 10x10 board
func DecodeBoard(data []byte, player1, player2 *Player) (*Board, map[Position]bool) {
	board := NewBoard()
	movedPositions := make(map[Position]bool)

	for y := range 10 {
		for x := range 10 {
			cell := data[y*10+x]
			if cell&BitOccupied == 0 {
				continue
			}

			pieceID := (cell & MaskPieceType) >> ShiftPieceType
			pieceType := idToPieceType[pieceID]

			owner := player1
			if cell&BitColor != 0 {
				owner = player2
			}

			board.SetPieceAt(NewPosition(x, y), NewPiece(*pieceType, owner))

			if cell&BitMoved != 0 {
				movedPositions[NewPosition(x, y)] = true
			}
		}
	}

	return board, movedPositions
}

// Encode to base64 string
func EncodeBoardToBase64(board *Board, movedPositions map[Position]bool) string {
	return base64.StdEncoding.EncodeToString(EncodeBoard(board, movedPositions))
}

// Decode from base64 string
func DecodeBoardFromBase64(encoded string, player1, player2 *Player) (*Board, map[Position]bool, error) {
	data, err := base64.StdEncoding.DecodeString(encoded)
	if err != nil {
		return nil, nil, err
	}
	if len(data) != 100 {
		return nil, nil, fmt.Errorf("invalid data length: expected 100, got %d", len(data))
	}
	board, moved := DecodeBoard(data, player1, player2)
	return board, moved, nil
}

// Encode 4x10 setup to 40 bytes
func EncodeSetup(rows []string, playerID int) ([]byte, error) {
	if len(rows) != 4 {
		return nil, fmt.Errorf("expected 4 rows, got %d", len(rows))
	}

	data := make([]byte, 40)
	for i, row := range rows {
		if len(row) != 10 {
			return nil, fmt.Errorf("row %d must be 10 chars, got %d", i, len(row))
		}

		for j, char := range row {
			if char == '.' || char == ' ' {
				continue
			}

			cell := byte(BitOccupied)
			cell |= (rankToPieceID[byte(char)] << ShiftPieceType) & MaskPieceType

			if playerID == 2 {
				cell |= BitColor
			}

			data[i*10+j] = cell
		}
	}

	return data, nil
}

// Decode 40 bytes to 4x10 setup
func DecodeSetup(data []byte) ([]string, int, error) {
	if len(data) != 40 {
		return nil, 0, fmt.Errorf("invalid data length: expected 40, got %d", len(data))
	}

	rows := make([]string, 4)
	playerID := 1

	for i := 0; i < 4; i++ {
		row := make([]byte, 10)
		for j := 0; j < 10; j++ {
			cell := data[i*10+j]
			if cell&BitOccupied == 0 {
				row[j] = '.'
			} else {
				pieceID := (cell & MaskPieceType) >> ShiftPieceType
				row[j] = pieceIDToRank[pieceID]

				if cell&BitColor != 0 {
					playerID = 2
				}
			}
		}
		rows[i] = string(row)
	}

	return rows, playerID, nil
}

// Validate setup has correct piece counts
func ValidateSetup(rows []string) error {
	if len(rows) != 4 {
		return fmt.Errorf("setup must have 4 rows, got %d", len(rows))
	}

	counts := make(map[byte]int)
	for _, row := range rows {
		if len(row) != 10 {
			return fmt.Errorf("each row must be 10 chars")
		}
		for _, c := range row {
			if c != '.' && c != ' ' {
				counts[byte(c)]++
			}
		}
	}

	expected := map[byte]int{
		'0': 1, 'B': 6, '1': 1, '2': 8, '3': 5, '4': 4,
		'5': 4, '6': 4, '7': 3, '8': 2, '9': 1, 'M': 1,
	}

	for piece, count := range expected {
		if counts[piece] != count {
			return fmt.Errorf("piece %c: expected %d, got %d", piece, count, counts[piece])
		}
	}

	return nil
}

File: code/frontend/src/lib/boardBinary.ts
/**
 * Binary format for Stratego board encoding
 * 8 bits per square: Occupied | Piece type | Color | Moved
 */

export const BitOccupied = 0x80;
export const BitColor = 0x02;
export const BitMoved = 0x01;
export const MaskPieceType = 0x7C;
export const ShiftPieceType = 2;

export enum PieceID {
	Flag = 1,
	Bomb = 2,
	Spy = 3,
	Scout = 4,
	Miner = 5,
	Sergeant = 6,
	Lieutenant = 7,
	Captain = 8,
	Major = 9,
	Colonel = 10,
	General = 11,
	Marshal = 12
}

const rankToPieceID: Record<string, PieceID> = {
	'0': PieceID.Flag, 'B': PieceID.Bomb, '1': PieceID.Spy, '2': PieceID.Scout,
	'3': PieceID.Miner, '4': PieceID.Sergeant, '5': PieceID.Lieutenant, '6': PieceID.Captain,
	'7': PieceID.Major, '8': PieceID.Colonel, '9': PieceID.General, 'M': PieceID.Marshal
};

const pieceIDToRank: Record<number, string> = Object.fromEntries(
	Object.entries(rankToPieceID).map(([k, v]) => [v, k])
);

export interface BoardSquare {
	occupied: boolean;
	pieceType?: PieceID;
	color?: number;
	moved?: boolean;
}

// Encode 10x10 board to 100 bytes
export function encodeBoard(board: BoardSquare[][]): Uint8Array {
	const data = new Uint8Array(100);
	
	for (let y = 0; y < 10; y++) {
		for (let x = 0; x < 10; x++) {
			const square = board[y][x];
			if (!square.occupied || !square.pieceType) continue;
			
			let cell = BitOccupied;
			cell |= (square.pieceType << ShiftPieceType) & MaskPieceType;
			
			if (square.color === 2) {
				cell |= BitColor;
			}
			
			if (square.moved) {
				cell |= BitMoved;
			}
			
			data[y * 10 + x] = cell;
		}
	}
	
	return data;
}

// Decode 100 bytes to 10x10 board
export function decodeBoard(data: Uint8Array): BoardSquare[][] {
	const board: BoardSquare[][] = [];
	
	for (let y = 0; y < 10; y++) {
		const row: BoardSquare[] = [];
		for (let x = 0; x < 10; x++) {
			const cell = data[y * 10 + x];
			
			if (cell & BitOccupied) {
				const pieceType = (cell & MaskPieceType) >> ShiftPieceType;
				row.push({
					occupied: true,
					pieceType,
					color: (cell & BitColor) ? 2 : 1,
					moved: !!(cell & BitMoved)
				});
			} else {
				row.push({ occupied: false });
			}
		}
		board.push(row);
	}
	
	return board;
}

// Encode 4x10 setup to 40 bytes
export function encodeSetup(rows: string[], playerID: number): Uint8Array {
	if (rows.length !== 4) throw new Error(`Expected 4 rows, got ${rows.length}`);
	
	const data = new Uint8Array(40);
	
	for (let i = 0; i < 4; i++) {
		if (rows[i].length !== 10) throw new Error(`Row ${i} must be 10 chars`);
		
		for (let j = 0; j < 10; j++) {
			const char = rows[i][j];
			if (char === '.' || char === ' ') continue;
			
			let cell = BitOccupied;
			cell |= (rankToPieceID[char] << ShiftPieceType) & MaskPieceType;
			
			if (playerID === 2) {
				cell |= BitColor;
			}
			
			data[i * 10 + j] = cell;
		}
	}
	
	return data;
}

// Decode 40 bytes to 4x10 setup
export function decodeSetup(data: Uint8Array): { rows: string[]; playerID: number } {
	if (data.length !== 40) throw new Error(`Invalid data length: expected 40, got ${data.length}`);
	
	const rows: string[] = [];
	let playerID = 1;
	
	for (let i = 0; i < 4; i++) {
		let row = '';
		for (let j = 0; j < 10; j++) {
			const cell = data[i * 10 + j];
			
			if (cell & BitOccupied) {
				const pieceID = (cell & MaskPieceType) >> ShiftPieceType;
				row += pieceIDToRank[pieceID];
				
				if (cell & BitColor) {
					playerID = 2;
				}
			} else {
				row += '.';
			}
		}
		rows.push(row);
	}
	
	return { rows, playerID };
}

// Validate setup piece counts
export function validateSetup(rows: string[]): { valid: boolean; error?: string } {
	if (rows.length !== 4) {
		return { valid: false, error: 'Setup must have 4 rows' };
	}
	
	const counts: Record<string, number> = {};
	for (const row of rows) {
		if (row.length !== 10) {
			return { valid: false, error: 'Each row must be 10 chars' };
		}
		for (const c of row) {
			if (c !== '.' && c !== ' ') {
				counts[c] = (counts[c] || 0) + 1;
			}
		}
	}
	
	const expected: Record<string, number> = {
		'0': 1, 'B': 6, '1': 1, '2': 8, '3': 5, '4': 4,
		'5': 4, '6': 4, '7': 3, '8': 2, '9': 1, 'M': 1
	};
	
	for (const [piece, count] of Object.entries(expected)) {
		if (counts[piece] !== count) {
			return {
				valid: false,
				error: `Piece ${piece}: expected ${count}, got ${counts[piece] || 0}`
			};
		}
	}
	
	return { valid: true };
}

// Helper: Create empty board
export function createEmptyBoard(): BoardSquare[][] {
	return Array.from({ length: 10 }, () =>
		Array.from({ length: 10 }, () => ({ occupied: false }))
	);
}

// Helper: Create empty setup
export function createEmptySetup(): string[] {
	return Array(4).fill('.'.repeat(10));
}
